Update 001-RPC-Interface-Specification.md (#5)
hyperledger-labs/perun-proposals: design/001-RPC-Interface-Specification.md

> !-- This is a template for proposing design changes to the perun project. -->
> 
> # Proposal: Descriptive specification of the User API for Two Party Payment Channels

Reviewer: ```suggestion
# Proposal: Descriptive specification of the User API for Two Party Payment Channels
```
I'd remove the template comments.

Author: Updated.

1. Node APIs - For accessing the node related functionality. Following APIs
   should be provided:

    1. [Get config](#1-get-config)
    2. [Open Session](#2-open-session)
    3. [Time](#3-time)
    4. [Help](#4-help)

2. Session APIs - For accessing the session related functionality. Each request
   should include a Session ID. Following APIs should be provided:

    1. [Add Contact](#1-add-contact)
    2. [Get Contact](#2-get-contact)
    3. [Open Payment Channel](#3-open-payment-channel)
    4. [Get Payment Channels Info](#4-get-payment-channels-info)
    5. [Subscribe To Payment Channel Proposals](#5-subscribe-to-payment-channel-proposals)

>     6. [Unsubscribe To Payment Channel Proposals](#6-unsubscribe-to-payment-channel-proposals)

Reviewer: ```suggestion
    6. [Unsubscribe From Payment Channel Proposals](#6-unsubscribe-from-payment-channel-proposals)
```

Author: Updated.

    7. [Respond To Payment Channel Proposal](#7-respond-to-payment-channel-proposal)
    8. [Close Session](#8-close-session)

> 3. Channel APIS - For accessing the channel related functionality. Each request

Reviewer: ```suggestion
3. Channel APIs - For accessing the channel related functionality. Each request
```

Author: Updated.

   should include a Session ID & Channel ID. Following APIs should be provided:

    1. [Send Payment Channel Update](#1-send-payment-channel-update)
    2. [Subscribe To Payment Channel Updates](#2-subscribe-to-payment-channel-updates)

>     3. [Unsubscribe To Payment Channel Updates](#3-unsubscribe-to-payment-channel-updates)

Reviewer: ```suggestion
    3. [Unsubscribe From Payment Channel Updates](#3-unsubscribe-from-payment-channel-updates)
```

    4. [Respond To Payment Channel Update](#4-respond-to-payment-channel-update)
    5. [Get Payment Channel Info](#5-get-payment-channel-info)
    6. [Close Payment Channel](#6-close-payment-channel)

### Data formats

The following 3 data formats will be used in the APIs.

#### 1. Peer

* `Alias`: [String] Alias for the peer. This will be used to reference a
  peer in all API calls.
* `Off-Chain Address`: [String] Permanent address (cryptographic) of the
  peer in the off-chain network.
* `Comm Address`: [String] Address (network) of the peer for off-chain
  communication.
* `Comm Type`: [String] Type of communication protocol supported by the
  peer for off-chain communication.

#### 2. Balance Info

* `Currency`: [String] Currency used for specifying the balance.
* `Aliases`: [List of String] Alias of each channel participants, as in the
  user's contacts provider. `self` is a special alias for the user of the
  node. Each entry in the list should be unique.
* `Balance`: [List of String] Amount held by each channel participant, in

>   the same order as that in the `Aliases`. Each entry will be a non
>   negative value and the length of balance list should be same as that of
>   the Aliases list.

Reviewer: ```suggestion
  the same order as that in the `Aliases`. Each entry will be a nonnegative
  value and the length of balance list should be same as that of
  the Aliases list.
```

Author: Updated.

* When using the `Balance Info` as a parameter in API calls, an error will
  be returned if:

>     * the peer corresponding to any of the Alias cannot be found in user's

Reviewer: ```suggestion
    * the peer corresponding to any of the Aliases cannot be found in user's
```

Author: Updated.

      contacts provider.
    * alias list does not have an entry `self` corresponding to the user.

>     * alias list has non unique entries.

Reviewer: or formulated positively,
```suggestion
    * alias list has duplicate entries.
```

Author: Updated.

    * `Currency`: [String] Currency used for specifying the amount.
    * `Balance`: [Map of `Alias` to `Balance`]
      * `Alias`: [String] `Self` is a special alias for the user of the node.

>       * `Balances`: [String] Balance of the `Peer` corresponding to the Alias.

Reviewer: This is unclear to me. Is the map's name supposed to be `Balances` and should the string representing a `Peer`'s balance be called `Balance` rather than the other way around?

Author: You are correct, it is more clear if map's name is `Balances` and the string field name is `Balance`. 
Updated.

Author: I have updated the format of payment channel info to use lists instead of maps. This has two advantages:
1. We require one term less. Instead of using `Balances` for the map and `Balance` for the field in previous version, now there is just one term `Balances` for the list.
2. When we add support for multiple assets (like other ERC20 tokens) to be held in the same channel, this structure can be scaled without repeating the list of aliases for each asset.

Reviewer: I'd propose to use the singular 'Balance' for the list of amounts held by the channel participants, as the whole list is the channel's balance (invariant: sum of amounts).
Additionally, the constraints regarding the Alias and Balance lists should be documented here (lists have same length, values in Alias are unique, Alias is a known peer, and amount held by a particular peer whose Alias is Alias[idx] is balance[idx], amounts in balance are non-negative, ...)
Please add the rationale why you don't choose a map data structure mapping Alias to Amount but two lists.

Author: Okay, I have updated the spec to use the term `Balance` and have also documented the constraints and rationale for using two lists instead of a map.

3. `Payment Channel Info`

    * `Channel ID`: [String] Unique ID of the channel.
    * `BalanceInfo`: [Balance Info]

>     * `Version`: [String] Current Version of the channel.

Reviewer: What's versino? Version of the software? Do you expect the node or session to support different versions of the protocol? Otherwise version query could be moved to either node or session API?

Author: It is the version of the state channel, which is incremented each time when the state is updated.

Reviewer 2: Since the term 'version' makes confusion with other versions like software version, protocol version etc, may be we can use the term 'stateCounter' instead of 'version' which actually increments the counter for each updation of state.

Author: Okay, now i get the point. I agree `StateCounter` is a more clear term than `Version`. Will update the spec.

Reviewer: Well, if the term `Version` is used for this field in go-perun, it may be better to keep in sync with the terminology for the time being. In that case I'd suggest to add a description of what it means.

Author: Okay, I have retained the term `Version` and have added additional description to it.

*Parameters*

* `Session ID`: [String] Unique ID of the session.

> * `PeerAlias`: [String] Alias of peer with whom channel should be opened.

Reviewer: Since the `Alias` parameter has been removed, I wonder which peer the payment channel is opened with?

Author: The `Balances` field will already contain the `Alias`es of all all the participants to whom channel is to be opened with. Hence I felt, this field was redundant.

Author: I have updated the format of payment channel info to use lists instead of maps. This has two advantages:
1. We require one term less. Instead of using `Balances` for the map and `Balance` for the field in previous version, now there is just one term `Balances` for the list.
2. When we add support for multiple assets (like other ERC20 tokens) to be held in the same channel, this structure can be scaled without repeating the list of aliases for each asset.

Subscribe to notifications on new incoming payment channel proposals in the
specified session.

Only one subscription can be made at a time. Making a new subscription without
canceling the previous one will return an error.

The incoming channel proposal received when there was no subscription will
have been cached by the node. Once a new subscription is made, node will send
these cached requests (if any), as individual notifications. It will then
continue to send a notification for each new incoming channel proposal.

Response to the notifications can be sent using the

> [`Respond-To Payment Channel Proposal`](#7-respond-to-payment-channel-proposal)

Reviewer: ```suggestion
[`Respond To Payment Channel Proposal`](#7-respond-to-payment-channel-proposal)
```

Author: Updated.

Each notification sent to the user should contain the following data:

* `Proposal ID`: [String] Unique ID of this channel proposal.
* `Opening Balance`: [[`Balance Info`](#2-balance-info)]
* `Challenge Duration in Seconds`: [uint64]
* `Expiry`: [int64] Time (in unix format) before which response should be sent.

> #### 6. Unsubscribe To Payment Channel Proposals

Reviewer: ```suggestion
#### 6. Unsubscribe From Payment Channel Proposals
```

Author: Updated.
